### PR TITLE
Add docs build for ReadTheDocs

### PR DESCRIPTION
**Pull request by bollwyvl**
[Binder](https://mybinder.org/v2/gh/bollwyvl/jupyterlite/add-docs-build?urlpath=lab)

## References

- starts #63

## Code changes

- [x] add `.readthedocs.yml`
  - [x] enabled  fail on warning, otherwise they stack up until it stops working
  - [x] sync the environment to binder 
- [x] Adds a sphinx site
  - [x] conf.py
  - [x] start some kernels stuff
    - [ ] couldn't figure out how to get e.g. `sphinx-autodoc` to work against pyolite... this can be a separate issue
- [x] adds a `dodo.py`
- [x] update contributing
- [x] remove checked-in `whl` 

## User-facing changes

> n/a

## Backwards-incompatible changes

- moved `pyolite` down one folder, since we might end up with `traitlites`, `litedgets`

### CODE REVIEW

**Comment by bollwyvl** [comment]
Guess we'll see if we get past `conda`... and hopefully don't get destroyed by webpack...

**Comment by bollwyvl** [comment]
Welp, we gotta site: https://jupyterlite--64.org.readthedocs.build/en/64/

**Comment by bollwyvl** [comment]
Looks like i messed up some URL stuff:
```
PythonError: Traceback (most recent call last):
  File "/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
  File "/lib/python3.8/asyncio/tasks.py", line 282, in __step
    result = coro.throw(exc)
  File "/lib/python3.8/site-packages/pyodide/_base.py", line 419, in eval_code_async
    return await CodeRunner(
  File "/lib/python3.8/site-packages/pyodide/_base.py", line 272, in run_async
    await coro
  File "<exec>", line 3, in <module>
  File "/lib/python3.8/asyncio/futures.py", line 260, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.8/asyncio/tasks.py", line 349, in __wakeup
    future.result()
  File "/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
  File "/lib/python3.8/asyncio/tasks.py", line 282, in __step
    result = coro.throw(exc)
  File "/lib/python3.8/site-packages/micropip.py", line 178, in install
    await gather(*wheel_promises)
  File "/lib/python3.8/asyncio/futures.py", line 260, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.8/asyncio/tasks.py", line 349, in __wakeup
    future.result()
  File "/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
  File "/lib/python3.8/asyncio/tasks.py", line 280, in __step
    result = coro.send(None)
  File "/lib/python3.8/site-packages/micropip.py", line 123, in _install_wheel
    wheel = await _get_url(url)
  File "/lib/python3.8/site-packages/micropip.py", line 35, in _get_url
    raise OSError(
OSError: Request for https://jupyterlite--64.org.readthedocs.build/en/64/_static/en/64/_static/lab/build/pyolite-0.1.0-py3-none-any.whl failed with status 404: Not Found
```

**Comment by bollwyvl** [comment]
On the JS tests, we should add coverage, or xunit, or something, so that we know it has run. Similarly, it would be ideal if webpack generated a list of all the files it emitted (or just a hash of them) somewhere. Otherwise the task entropy quacks pretty good.

**Comment by bollwyvl** [comment]
So the first one exists (i suspected RTD clobbering wheels):
```
https://jupyterlite--64.org.readthedocs.build/en/64/_static/lab/build/pyolite-0.1.0-py3-none-any.whl
https://jupyterlite--64.org.readthedocs.build/en/64/_static/en/64/_static/lab/build/pyolite-0.1.0-py3-none-any.whl 
```
We're just getting some extra `/` in there somehow...

**Comment by bollwyvl** [comment]
Seems to be deploying correctly:

Screenshot from 2021-04-24 22-06-47

It's super funny that the notebooks persist in localstorage during the course of a PR.

**Review by jtpio — APPROVED** [review]
Wow, thanks @bollwyvl for this!

**Comment by jtpio** [comment]
> couldn't figure out how to get e.g. sphinx-autodoc to work against pyolite... this can be a separate issue

No problem this can be tracked separately.

> It's super funny that the notebooks persist in localstorage during the course of a PR.

Right, probably because the domain stays the same for the PR (`jupyterlite--64.org.readthedocs.build` for that one).

File: docs/kernels/index.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+JupyterLite Kernels implement [Jupyter Kernel Messaging][jkm] in the browser with the
+help of [`mock-socket`][mock-socket].
+
+| Feature       | Message                                       | [pyolite](./pyolite.md) | [javascript](./js.md) | [p5](./p5.md) |
+| ------------- | --------------------------------------------- | ----------------------- | --------------------- | ------------- |
+|               | Language                                      | Python 3.8              | Browser JS            | JS + p5.js    |
+| Start session | `kernel_info_request`<br/>`kernel_info_reply` | ✔️                      | ✔️                    | ✔️            | ✔️ |
+| Run code      | `execute_request`<br/>`execute_reply`         | ✔️                      | ✔️                    | ✔️            | ✔️ |
+| Completion    | `complete_request`<br/>`complete_reply`       | ✔️                      | ✔️                    | ✔️            |
```
Comment (jtpio) [inline docs/kernels/index.md:11]: Technically we don't really have completion yet for Pyolite, and the one for JS is really basic.

**Comment by jtpio** [comment]
Also Vercel should now be disconnected from the repo.

Later on, it would still be interesting to try uploading the dist with the GitHub Action (as mentioned in https://github.com/jtpio/jupyterlite/issues/63). This could also serve as documentation for deploying JupyterLite.

**Comment by jtpio** [comment]
> On the JS tests, we should add coverage, or xunit, or something, so that we know it has run. Similarly, it would be ideal if webpack generated a list of all the files it emitted (or just a hash of them) somewhere. Otherwise the task entropy quacks pretty good.

Absolutely. And also a quick end to end (#6) test to check both lab and classic are up and running.

Just realized that #62 might have broken classic (or at least the `/tree` redirect) which seems to be the case on Binder or when serving locally with `yarn serve`:

image

**Comment by bollwyvl** [comment]
I think it's probably just the tree redirect... it doesn't know what to do with an un-trailing-slashed file (there could be a file called `tree` without an extension, i guess). Will look into it!

**Comment by jtpio** [comment]
Thanks!

**Comment by bollwyvl** [comment]
Cool. I'll make some additional follow-on issues.

**Comment by bollwyvl** [comment]
We should probably set up a redirect on e.g /demo or /try so that the link
can be shorter.

**Pull request merged**